### DISCUSSION TITLE
Help: native client app use?

### DISCUSSION THREAD

**Discussion by shyfishy** [zitadel/oidc#628] [Q&A] [answered]
Hi, I'm a bit unsure how to go about it when it's a native client app and service + auth-server (rather than just a webapp and auth-server). I think in that case the `example/client/app` example is analogous to the service (it's what handles the callback request?) rather than the native client app. How should the native client app get the access token if it's the service that gets the token in the callback?
Thank you!

**Comment by muhlemmer** [comment, marked as answer]
For a native app, you can start a HTTP listener on the localhost, with a random port and a callback path. It the Provider sees a client type `native`, it allows redirects to random port numbers. So the native app will need to:

1. Start a HTTP server on a random / unprivileged port
2. Form a redirect URL `http://localhost:<port>/callback`
3. Build the auth request URL with the required parameters including the `redirect_url`
4. Open the OS' browser point to the auth request URL.
   - On linux systems the `xdg-open` mechanism can be used for example
   - If you use Go, [`browser.OpenURL()`](https://pkg.go.dev/github.com/pkg/browser#OpenURL) might be a good option
5. User completes login flow
6. User is redirected with authorization code to the `redirect_url` created above
7. Native app calls token endpoint to obtain tokens.
8. The HTTP server can be terminated.

Our client packages and examples do not have native app support yet. I did start working on an example for my own fun and education, but as demand for native apps in Go is low I don't dedicate much time on it.

**Reply by shyfishy** [reply]
Thank you so much for laying it out for me! This is great!
